frontend/components/generateProof.tsx

```diff
@@ -0,0 +1,199 @@
+      const signatureArtifacts: SignaturePostProcessingContents = {
+        TPreComputes,
+        s: splitToRegisters(s.substring(2)) as bigint[],
+        U: [
+          splitToRegisters(U.x) as bigint[],
+          splitToRegisters(U.y) as bigint[],
+        ],
+      };
+      await generateProof(signatureArtifacts);
+    },
+  });
+
+  const generateProof = React.useCallback(
+    async (artifacts: SignaturePostProcessingContents) => {
+      if (!merkleTreeProofData.current) {
+        throw new Error("Missing merkle tree data");
+      }
+      //TODO: add loading state or progress bar first time it downloads zkey
+      await downloadZKey();
+
+      const proofInputs = { ...artifacts, ...merkleTreeProofData.current };
+      const zkeyDb = await localforage.getItem("setMembership_final.zkey");
+
+      if (!zkeyDb) {
+        throw new Error("zkey was not found in the database");
+        return;
+      }
+
+      // @ts-ignore
+      const zkeyRawData = new Uint8Array(zkeyDb);
+
+      const zkeyFastFile = { type: "mem", data: zkeyRawData };
+      const worker = new Worker("./worker.js");
+      worker.postMessage([proofInputs, zkeyFastFile]);
+      worker.onmessage = async function (e) {
+        const { proof, publicSignals } = e.data;
+        console.log("PROOF SUCCESSFULLY GENERATED: ", proof, publicSignals);
+        // TODO: post to IPFS or store in our db
+        setSuccessProofGen(true);
+        // TODO: add toast showing success
+        setLoadingText(undefined);
+      };
+    },
+    []
+  );
+
+  const prepareProof = React.useCallback(async () => {
+    try {
+      const merkleTreeData = (
+        await axios.get("/api/getPropGroup", {
+          params: {
+            userAddr: address,
+            propId: propId,
+            groupId: groupId,
+          },
+        })
+      ).data;
+
+      // TODO: REMOVE THIS AFTER TESTING, generating dummy merkle tree to test proof generation works
+      //       if you want to test non-noun holding addresses
+      // const { pathElements, pathIndices, pathRoot } = await createMerkleTree(
+      //   "0x926B47C42Ce6BC92242c080CF8fAFEd34a164017",
+      //   [
+      //     "0x926B47C42Ce6BC92242c080CF8fAFEd34a164017",
+      //     "0x199D5ED7F45F4eE35960cF22EAde2076e95B253F",
+      //   ]
+      // );
+
+      // const merkleTreeData = prepareMerkleRootProof(
+      //   pathElements,
+      //   pathIndices,
+      //   pathRoot
+      // );
+
+      merkleTreeProofData.current = merkleTreeData;
+
+      const groupMessage =
+        groupId === 1
+          ? "I own one Noun"
+          : groupId === 2
+          ? "I own two or more Nouns"
+          : "I am a Nounder";
+      // triggers callback which will call generateProof when it's done
+      signMessage({
+        message: `As of Prop ${propId}, I attest that ${groupMessage} and the hash (keccak) of the message I'm posting is ${ethers.utils.hashMessage(
```

Comment:
nice!

Reply:
btw @amirgamil is it easy to do eip712 with wagmi? this has the added benefit of making it so that we can add structured data to what's visible in the web3 wallet to sign, like:
```
{
propId: N,
groupMessage: Y,
...
}
```
that might be easier for a user to parse. 

we did this for heyanon: 
https://github.com/personaelabs/heyanon/blob/main/pages/postmsg/%5BgroupId%5D.tsx#L135
https://github.com/personaelabs/heyanon/blob/main/lib/hashing.ts#L33